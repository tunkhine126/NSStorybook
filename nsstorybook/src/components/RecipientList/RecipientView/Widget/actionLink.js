/**
 *
 * ActionLink
 *
 */

import React from 'react';
import PropTypes from 'prop-types';
import { FormattedMessage } from 'react-intl';

import TextLink from 'components/shared/TextLink';
import globalMessages from 'messages';
import { recipientListStyles } from '../../styles';

function ActionLink({ message, ...rest }) {
  const classes = recipientListStyles();

  return (
    <TextLink
      {...rest}
      className={classes.addBtn}
      content={<FormattedMessage {...globalMessages[message]} />}
    />
  );
}

ActionLink.propTypes = {
  message: PropTypes.string.isRequired,
};

export default ActionLink;
