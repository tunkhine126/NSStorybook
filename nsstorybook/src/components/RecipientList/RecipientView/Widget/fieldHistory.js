/**
 *
 * FieldHistory
 *
 */

import React from 'react';
import PropTypes from 'prop-types';
import { injectIntl, intlShape } from 'react-intl';

import Grid from '@material-ui/core/Grid';
import Typography from '@material-ui/core/Typography';
import Divider from '@material-ui/core/Divider';

import globalMessages from 'messages';
import { FieldHistoryStyle } from '../../styles';

function FieldHistory({ intl, date, value, user, currentVersion, lastItem }) {
  const classes = FieldHistoryStyle();

  return (
    <>
      <Grid
        item
        container
        className={classes.container}
        direction="column"
        justify="center"
        alignItems="flex-start"
      >
        {currentVersion && (
          <Typography variant="body1" className={classes.currentVersion}>
            {intl.formatMessage(globalMessages.currentVersion)}
          </Typography>
        )}

        <Typography variant="body1" className={classes.user}>
          {date} â€“ {user}
        </Typography>
        <Typography variant="body1" className={classes.content}>
          {value}
        </Typography>
      </Grid>

      {!lastItem && (
        <Grid
          item
          container
          direction="row"
          justify="center"
          alignItems="flex-start"
          className={classes.lastItem}
        >
          <Divider light className={classes.divider} />
        </Grid>
      )}
    </>
  );
}

FieldHistory.propTypes = {
  intl: intlShape.isRequired,
  date: PropTypes.string.isRequired,
  value: PropTypes.string.isRequired,
  user: PropTypes.string.isRequired,
  currentVersion: PropTypes.bool,
  lastItem: PropTypes.bool,
};

const intlFieldHistory = injectIntl(FieldHistory);

export default intlFieldHistory;
