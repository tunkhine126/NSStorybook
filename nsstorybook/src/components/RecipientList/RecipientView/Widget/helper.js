/* eslint-disable no-restricted-syntax */
/* eslint-disable react/display-name */
import React from 'react';
import moment from 'moment';
import PropTypes from 'prop-types';

import InputValidation from 'components/shared/InputValidation';
import DatePickerValidation from 'components/shared/DatePickerValidation';
import TimePickerValidation from 'components/shared/TimePickerValidation';
import SelectValidation from 'components/shared/SelectValidation';
import MultiSelectValidation from 'components/shared/MultiSelectValidation';
import Photo from './fieldTypes/photo';

import { SelectWrapper } from '../../styles';
import { FieldsProvider } from '../context/fields';
import SwitchMode from './switchMode';

export const ftSwitch = (
  field,
  { customFields, uuid },
  key,
  rData,
  handleOptionChange,
  handleDateTimeChange,
  updateRecipient
) => {
  let optionValues = [];
  let optionIntialValue = {};
  let optionIsNull = false;
  const enLabel = field.contexts.us_en.label;

  if (field.option_values) {
    optionValues = field.option_values.map(i => ({ value: i }));

    if (!rData.currentValue[field.name].item) {
      optionIsNull = true;
    }

    optionIntialValue = !optionIsNull
      ? rData.currentValue[field.name].item
      : optionValues[0].value;
  }

  const Field = ({
    classes,
    label,
    name,
    content,
    element,
    fieldType,
    type,
  }) => (
    <FieldsProvider type={type}>
      <SwitchMode
        classes={classes}
        rData={rData}
        field={field}
        uuid={uuid}
        fieldType={fieldType}
        handleDateTimeChange={handleDateTimeChange}
        handleOptionChange={handleOptionChange}
        updateRecipient={updateRecipient}
        label={label}
        currentField={name}
        content={content}
        element={element}
      />
    </FieldsProvider>
  );

  Field.propTypes = {
    classes: PropTypes.string.isRequired,
    label: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    element: PropTypes.any.isRequired,
    content: PropTypes.any,
    fieldType: PropTypes.string,
    type: PropTypes.string,
  };

  switch (field.field_type) {
    case 'short_text':
      return (
        <Field
          key={key}
          classes="field-container"
          label={enLabel}
          name={field.name}
          content={customFields[field.name]}
          element={
            <InputValidation
              maxWidth="none"
              field={field.name}
              type="text"
              initialValue={customFields[field.name]}
              label={enLabel}
            />
          }
        />
      );
    case 'paragraph_text':
      return (
        <Field
          key={key}
          classes="paragraph-text field-container"
          label={enLabel}
          name={field.name}
          content={customFields[field.name]}
          element={
            <InputValidation
              field={field.name}
              maxWidth="none"
              type="text"
              multiline
              rows="3"
              initialValue={customFields[field.name]}
              label={enLabel}
            />
          }
        />
      );
    case 'number':
      return (
        <Field
          key={key}
          classes="field-container"
          label={enLabel}
          name={field.name}
          content={
            customFields[field.name] ? customFields[field.name].toString() : ''
          }
          element={
            <InputValidation
              type="number"
              maxWidth="none"
              field={field.name}
              initialValue={
                customFields[field.name]
                  ? customFields[field.name].toString()
                  : ''
              }
              label={enLabel}
            />
          }
        />
      );
    case 'date':
      return (
        <Field
          key={key}
          classes="field-container"
          label={enLabel}
          name={field.name}
          content={rData.dates[field.name]}
          element={
            <DatePickerValidation
              field={field.name}
              label={enLabel}
              initialValue={rData.dates[field.name]}
              customFormat="MM-dd-yyyy"
              placeholder={`${moment().format('MM-DD-YYYY')}`}
              variant="dialog"
              inputVariant="filled"
              margin="normal"
              fullWidth
              clearable
              disableFuture
            />
          }
        />
      );
    case 'time':
      return (
        <Field
          key={key}
          classes="field-container"
          label={enLabel}
          name={field.name}
          content={
            rData.times[field.name]
              ? moment(new Date(rData.times[field.name])).format('hh:mm a')
              : null
          }
          fieldType="time"
          element={
            <TimePickerValidation
              field={field.name}
              label={enLabel}
              initialValue={rData.times[field.name]}
              placeholder="08:00 AM"
              inputVariant="filled"
              margin="normal"
              fullWidth
              clearable
            />
          }
        />
      );
    case 'single_select':
      return (
        <Field
          key={key}
          classes="field-container"
          label={enLabel}
          name={field.name}
          content={customFields[field.name]}
          element={(localChildData, localOptionChange) => (
            <SelectWrapper>
              <SelectValidation
                field={field.name}
                type="select"
                maxWidth="none"
                label={enLabel}
                values={
                  !optionIsNull
                    ? localChildData.currentValue[field.name]
                    : optionIntialValue
                }
                initialValue={optionIntialValue}
                items={optionValues}
                onChange={v => localOptionChange(v, field.name)}
                padding="25px 18px 10px 13px"
              />
            </SelectWrapper>
          )}
        />
      );
    case 'multi_select':
      return (
        <Field
          key={key}
          classes="field-container"
          label={enLabel}
          name={field.name}
          content={
            Array.isArray(customFields[field.name])
              ? customFields[field.name].join(', ')
              : customFields[field.name]
          }
          element={(localChildData, localOptionChange) => (
            <SelectWrapper>
              <MultiSelectValidation
                field={field.name}
                type="select"
                maxWidth="none"
                label={enLabel}
                values={
                  !optionIsNull
                    ? localChildData.currentValue[field.name]
                    : optionIntialValue
                }
                initialValue={optionIntialValue}
                onChange={v => localOptionChange(v, field.name)}
                items={optionValues}
                padding="25px 18px 10px 13px"
              />
            </SelectWrapper>
          )}
        />
      );
    case 'likert':
      return (
        <Field
          key={key}
          classes="field-container"
          label={enLabel}
          name={field.name}
          content={customFields[field.name]}
          element={(localChildData, localOptionChange) => (
            <SelectWrapper>
              <SelectValidation
                field={field.name}
                type="select"
                maxWidth="none"
                label={enLabel}
                values={
                  !optionIsNull
                    ? localChildData.currentValue[field.name]
                    : optionIntialValue
                }
                initialValue={optionIntialValue}
                items={optionValues}
                onChange={v => localOptionChange(v, field.name)}
                padding="25px 18px 10px 13px"
              />
            </SelectWrapper>
          )}
        />
      );
    case 'photo':
      return (
        <Field
          type="photo"
          key={key}
          classes="field-container"
          label={enLabel}
          name={field.name}
          content={
            <Photo
              field={field}
              label={enLabel}
              values={customFields}
              recipientData={rData}
            />
          }
          element={
            <Photo
              className="photo-edit-mode"
              rData={rData}
              editMode
              field={field}
              label={enLabel}
              values={customFields}
              recipientData={rData}
            />
          }
        />
      );
    case 'audio':
      return 'audio';
    case 'video':
      return 'video';
    case 'coordinates':
      return (
        <Field
          key={key}
          classes="field-container"
          label={enLabel}
          name={field.name}
          content={customFields[field.name]}
          element={
            <InputValidation
              maxWidth="none"
              field={field.name}
              type="text"
              initialValue={customFields[field.name]}
              label={enLabel}
            />
          }
        />
      );
    default:
      return null;
  }
};
