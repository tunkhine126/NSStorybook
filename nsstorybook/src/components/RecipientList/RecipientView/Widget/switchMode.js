/**
 *
 * SwitchMode
 *
 */

import React, { useState, useRef, useContext } from 'react';
import PropTypes from 'prop-types';
import { Form } from 'informed';
import moment from 'moment';
import { injectIntl, intlShape } from 'react-intl';

import Grid from '@material-ui/core/Grid';

import NSButton from 'components/shared/NSButton';
import TextLink from 'components/shared/TextLink';
import globalMessages from 'messages';
import { RecipientViewContext } from '../context';
import { SwithModeStyle } from '../../styles';
import ColumnContent from './columnContent';
import { FieldsContext } from '../context/fields';

function SwitchMode({
  intl,
  element,
  label,
  content,
  classes,
  currentField,
  fieldType,
  rData,
  field,
  uuid,
  handleDateTimeChange,
  handleOptionChange,
  updateRecipient,
}) {
  const localClasses = SwithModeStyle();
  const [editMode, setEditMode] = useState(false);
  const formApiRef = useRef();
  const formattedName = label.toLowerCase().replace(/ /g, '');
  const {
    toasterProps: { setToaster },
    saveProps: { setSaveState },
  } = useContext(RecipientViewContext);
  const {
    recipientPhotoProps: { recipientPhoto },
  } = useContext(FieldsContext);

  const [localChildData, setLocalChildData] = useState(rData);

  const localOptionChange = (value, fieldToChange) => {
    setLocalChildData(state => ({
      ...state,
      currentValue: {
        ...state.currentValue,
        [fieldToChange]: { item: value },
      },
    }));
  };

  const responseHandler = (response, fieldToSubmit) => {
    if (response.data) {
      let undoSettings;
      // Open Toaster
      if (field.field_type === 'photo') {
        undoSettings = () => handleSave({ ...rData.photos }, true);
      } else if (field.field_type === 'time') {
        undoSettings = () =>
          handleSave({
            [currentField]: rData.times[field.name],
          });
      } else if (field.field_type === 'multi_select') {
        undoSettings = () =>
          handleSave({
            [currentField]: rData.currentValue[currentField].item,
          });
      } else {
        undoSettings = () => handleSave({ [currentField]: content });
      }

      setToaster({
        show: true,
        recipientUuid: uuid,
        field: currentField,
        undoHandler: () => undoSettings(),
      });

      if (field.field_type === 'date') {
        if (moment(fieldToSubmit[currentField]).isValid()) {
          handleDateTimeChange(
            moment(fieldToSubmit[currentField]).format('YYYY-MM-DD'),
            field.name,
            'dates'
          );
        } else {
          handleDateTimeChange(null, field.name, 'dates');
        }
      }

      if (field.field_type === 'time') {
        if (fieldToSubmit[currentField]) {
          handleDateTimeChange(
            moment(new Date(fieldToSubmit[currentField])).toLocaleString(),
            field.name,
            'times'
          );
        } else {
          handleDateTimeChange(null, field.name, 'times');
        }
      }

      if (
        field.field_type === 'single_select' ||
        field.field_type === 'multi_select' ||
        field.field_type === 'likert'
      ) {
        handleOptionChange(fieldToSubmit[currentField], currentField);
      }
    }
  };

  const handleSave = async (data, undoPhoto) => {
    // convert empty strings to null
    let mergedData;
    const scrubbedData = { ...data };
    const entries = Object.entries(data);

    entries.forEach(e => {
      if (e[1] === '' || e[1] === 'Invalid date') {
        scrubbedData[e[0]] = null;
      } else if (fieldType === 'time' && !data[e[0]]) {
        scrubbedData[e[0]] = '';
      } else {
        scrubbedData[e[0]] = data[e[0]];
      }
    });

    if (!scrubbedData[currentField]) {
      scrubbedData[currentField] = null;
    }

    if (undoPhoto) {
      mergedData = { ...data };
    } else {
      mergedData = { ...scrubbedData, ...recipientPhoto };
    }

    const fieldToSubmit = { [currentField]: mergedData[currentField] };

    // START LOADING
    setSaveState({
      isLoading: true,
      recipientUuid: uuid,
      field: currentField,
    });

    try {
      const response = await updateRecipient({
        variables: {
          uuid,
          customFields: JSON.stringify(fieldToSubmit),
        },
      });

      responseHandler(response, fieldToSubmit);
    } catch (error) {
      console.log('ERROR', error);
    }
    // STOP LOADING
    setSaveState({ isLoading: false, recipientUuid: null, field: null });
  };

  return (
    <span className={classes}>
      {editMode ? (
        <Form
          getApi={formApi => {
            formApiRef.current = formApi;
          }}
          id={`${formattedName}-form`}
          onSubmit={d => {
            setEditMode(false);
            handleSave(d);
          }}
        >
          <div className={localClasses.formContent}>
            {field.field_type === 'single_select' ||
            field.field_type === 'multi_select' ||
            field.field_type === 'likert'
              ? element(localChildData, localOptionChange)
              : element}
            <Grid
              container
              direction="row"
              justify="flex-end"
              alignItems="center"
            >
              <TextLink
                className="edit-cancel-btn"
                content={intl.formatMessage(globalMessages.cancel)}
                onClick={() => setEditMode(false)}
              />
              <NSButton
                componentClasses="edit-save-btn"
                id="save-button"
                type="submit"
                version={2}
              >
                {intl.formatMessage(globalMessages.save)}
              </NSButton>
            </Grid>
          </div>
        </Form>
      ) : (
        <ColumnContent
          customClasses="recipient-column"
          header={label}
          handleEditMode={setEditMode}
          recipientUuid={uuid}
          field={currentField}
          data={content || ''}
          dynamic
        />
      )}
    </span>
  );
}

SwitchMode.propTypes = {
  intl: intlShape.isRequired,
  element: PropTypes.oneOfType([PropTypes.element, PropTypes.func]).isRequired,
  label: PropTypes.string.isRequired,
  currentField: PropTypes.string.isRequired,
  rData: PropTypes.any.isRequired,
  field: PropTypes.object.isRequired,
  uuid: PropTypes.string,
  handleDateTimeChange: PropTypes.func.isRequired,
  handleOptionChange: PropTypes.func.isRequired,
  updateRecipient: PropTypes.func.isRequired,
  content: PropTypes.any,
  classes: PropTypes.string,
  fieldType: PropTypes.string,
};

const intlSwitchMode = injectIntl(SwitchMode);

export default intlSwitchMode;
