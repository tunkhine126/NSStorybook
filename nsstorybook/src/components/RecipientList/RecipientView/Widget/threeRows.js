/**
 *
 * ThreeRows
 *
 */

import React from 'react';
import PropTypes from 'prop-types';
import clsx from 'clsx';

import Grid from '@material-ui/core/Grid';
import Typography from '@material-ui/core/Typography';

import { recipientListStyles, ThreeRowsStyle } from '../../styles';

function ThreeRows({ row1, row2, row3 }) {
  const classes = recipientListStyles();

  return (
    <Grid
      container
      direction="column"
      justify="flex-start"
      alignItems="flex-start"
    >
      <Grid item className={clsx(classes.captionSpacing)}>
        <Typography variant="body1" className={classes.threeRowTitle}>
          {row1}
        </Typography>
      </Grid>
      <Grid item>
        <Typography variant="body2" className={classes.contentBody}>
          <ThreeRowsStyle>{row2}</ThreeRowsStyle>
        </Typography>
      </Grid>
      <Grid item className={classes.columnWidth}>
        <Typography variant="caption" className={classes.caption}>
          {row3}
        </Typography>
      </Grid>
    </Grid>
  );
}

ThreeRows.propTypes = {
  row1: PropTypes.string.isRequired,
  row2: PropTypes.oneOfType([PropTypes.string, PropTypes.element]).isRequired,
  row3: PropTypes.string.isRequired,
};

export default ThreeRows;
