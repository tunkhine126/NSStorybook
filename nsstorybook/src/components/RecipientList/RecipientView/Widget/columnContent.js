/**
 *
 * ColumnContent
 *
 */

import React, { useContext, useState } from 'react';
import PropTypes from 'prop-types';
import clsx from 'clsx';
import { injectIntl, intlShape } from 'react-intl';

import Grid from '@material-ui/core/Grid';
import Typography from '@material-ui/core/Typography';
import EditIcon from '@material-ui/icons/Edit';
import AccessTimeIcon from '@material-ui/icons/AccessTime';
import CircularProgress from '@material-ui/core/CircularProgress';
import Modal from '@material-ui/core/Modal';
import Backdrop from '@material-ui/core/Backdrop';
import Fade from '@material-ui/core/Fade';

import NSButton from 'components/shared/NSButton';
import InlineToaster from 'components/shared/InlineToaster';
import moment from 'moment';
import globalMessage from 'messages';
import Widget from './widget';
import { recipientListStyles } from '../../styles';
import { RecipientViewContext } from '../context';
import FieldHistory from './fieldHistory';

function ColumnContent({
  intl,
  header,
  data,
  customClasses,
  dynamic,
  handleEditMode,
  recipientUuid,
  field,
  ...rest
}) {
  const classes = recipientListStyles();
  const {
    toasterProps: { toaster, setToaster },
    saveProps: { saveState },
  } = useContext(RecipientViewContext);

  const showToaster =
    toaster.show &&
    toaster.field === field &&
    toaster.recipientUuid === recipientUuid;

  const loading =
    saveState.isLoading &&
    saveState.recipientUuid === recipientUuid &&
    saveState.field === field;

  const [open, setOpen] = useState(false);

  const handleOpen = () => {
    setOpen(true);
  };

  const handleClose = () => {
    setOpen(false);
  };

  // TODO: Finish this feature
  const fieldHistory = [
    {
      timestamp: '2019-10-01T14:55:06.789Z',
      value: 'Eddie Dozier',
      user_uuid: '77c7702c-a4bc-11e9-bb06-22000a96ad48',
      user_name: 'Eddie Dozier',
    },
    {
      timestamp: '2019-10-02T14:28:19.535Z',
      value: 'Jimmy Hindrix',
      user_uuid: '77c7702c-a4bc-11e9-bb06-22000a96ad48',
      user_name: 'Eddie Dozier',
    },
    {
      timestamp: '2019-10-03T14:28:03.317Z',
      value: 'Maxwell',
      user_uuid: '77c7702c-a4bc-11e9-bb06-22000a96ad48',
      user_name: 'Eddie Dozier',
    },
  ].map(({ timestamp, value, user_name: userName }, idx, arr) => (
    <FieldHistory
      key={idx}
      date={moment(timestamp)
        .format('MMM DD, YYYY hh:mm:ss')
        .toLocaleString()}
      value={value}
      user={userName}
      currentVersion={idx === 0}
      lastItem={arr.length - 1 === idx}
      field={header}
    />
  ));

  return (
    <Grid
      {...rest}
      container
      className={customClasses}
      direction="row"
      justify="space-between"
      alignItems="center"
    >
      {loading ? (
        <Grid
          item
          container
          direction="column"
          justify="center"
          alignItems="center"
        >
          <CircularProgress />
        </Grid>
      ) : (
        <>
          <Grid item xs={9}>
            <Grid item className={clsx(classes.captionSpacing)}>
              <Typography variant="caption" className={classes.caption}>
                {header}
              </Typography>
            </Grid>

            <Grid item>
              {typeof data === 'string' ? (
                <Typography variant="body2" className={classes.contentBody}>
                  {data}
                </Typography>
              ) : (
                data
              )}
            </Grid>
          </Grid>
          {dynamic && (
            <Grid
              item
              xs={3}
              container
              direction="row"
              justify="flex-end"
              alignItems="center"
              className={clsx('column-action-btns', classes.columnActionBtns)}
            >
              {!toaster.show && (
                <Grid
                  item
                  container
                  direction="column"
                  justify="flex-end"
                  alignItems="center"
                >
                  <Grid item component="span" className={clsx(classes.editBtn)}>
                    <span className="edit-recipient">
                      <NSButton
                        version={2}
                        onClick={() => handleEditMode(true)}
                      >
                        <EditIcon />
                      </NSButton>
                    </span>
                  </Grid>

                  <Grid item component="span">
                    <span className="access-time">
                      <NSButton
                        onClick={handleOpen}
                        id={`${field}_${recipientUuid}`}
                      >
                        <AccessTimeIcon />
                      </NSButton>
                    </span>
                  </Grid>

                  <Modal
                    aria-labelledby={`${field}_${recipientUuid}`}
                    aria-describedby={`${field}_${recipientUuid}`}
                    open={open}
                    className={classes.modal}
                    onClose={handleClose}
                    closeAfterTransition
                    BackdropComponent={Backdrop}
                    BackdropProps={{
                      timeout: 400,
                    }}
                  >
                    <Fade in={open}>
                      <div className={classes.columnPaper}>
                        <Widget
                          header={`${intl.formatMessage(
                            globalMessage.fieldHistory
                          )}: ${header}`}
                          body={
                            <Grid
                              container
                              direction="row"
                              justify="center"
                              alignItems="flex-start"
                              className={classes.fieldHistory}
                            >
                              {fieldHistory}
                            </Grid>
                          }
                          message="closeWindow"
                          buttonPosition="end"
                          btnHandler={handleClose}
                        />
                      </div>
                    </Fade>
                  </Modal>
                </Grid>
              )}

              <Grid item>
                {showToaster && (
                  <InlineToaster
                    variant="success"
                    className="success-toaster"
                    message={intl.formatMessage(globalMessage.saved)}
                    showUndo
                    clearToaster={setToaster}
                    handleUndo={toaster.undoHandler}
                  />
                )}
              </Grid>
            </Grid>
          )}
        </>
      )}
    </Grid>
  );
}

ColumnContent.propTypes = {
  intl: intlShape.isRequired,
  header: PropTypes.oneOfType([PropTypes.string, PropTypes.element]).isRequired,
  data: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
    PropTypes.element,
  ]).isRequired,
  customClasses: PropTypes.string,
  field: PropTypes.string,
  recipientUuid: PropTypes.string,
  handleEditMode: PropTypes.func,
  dynamic: PropTypes.bool,
};

const intlColumnContent = injectIntl(ColumnContent);

export default intlColumnContent;
