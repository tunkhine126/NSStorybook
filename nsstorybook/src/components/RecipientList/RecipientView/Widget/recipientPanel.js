/**
 *
 * RecipientPanel
 *
 */

import React, { useState, useContext } from 'react';
import PropTypes from 'prop-types';
import { injectIntl, intlShape } from 'react-intl';
import { useMutation } from '@apollo/react-hooks';

import ExpansionPanel from '@material-ui/core/ExpansionPanel';
import ExpansionPanelSummary from '@material-ui/core/ExpansionPanelSummary';
import ExpansionPanelDetails from '@material-ui/core/ExpansionPanelDetails';
import Typography from '@material-ui/core/Typography';
import Grid from '@material-ui/core/Grid';
import CardMedia from '@material-ui/core/CardMedia';
import Divider from '@material-ui/core/Divider';
import DeleteIcon from '@material-ui/icons/Delete';

import { generalErrorHandler } from 'utils/error-handler';
import { UPDATE_RECIPIENT } from 'graphql/mutations/families/families-mutation';
import {
  GET_FAMILY_RECIEPIENT_CACHE,
  GET_FAMILY_CHILD_RECIEPIENT_CACHE,
} from 'graphql/queries/families/family-queries';
import SmallRightArrow from 'components/shared/Icons/smallRightArrow';
import NSButton from 'components/shared/NSButton';
import placeHolderImg from 'assets/images/familyPlaceholder.png';
import { generateImg } from 'utils/helpers';
import globalMessages from 'messages';
import TwoColumns from './twoColumns';
import { ftSwitch } from './helper';
import { RecipientViewContext } from '../context';
import {
  recipientListStyles,
  ExpansionPanelDetailsStyle,
  RecipientPanelStyles,
} from '../../styles';

function RecipientPanel({
  intl,
  name,
  type,
  data,
  customFields,
  enableDelete,
  parent = false,
  lastChild = true,
  twoColumn = false,
}) {
  let allOptions = [];
  const singleFields = [];
  let twoColfields = [];
  const allFieldsCurrentValue = {};
  const allPhotoFields = {};
  const allDateFields = {};
  const allTimeFields = {};
  const { rest, match } = useContext(RecipientViewContext);
  const classes = recipientListStyles();
  const rsClasses = RecipientPanelStyles();

  const [updateRecipientChild, { error: childError }] = useMutation(
    UPDATE_RECIPIENT,
    {
      update(
        cache,
        {
          data: {
            updateRecipient: {
              recipient: {
                parent: { children: updatedChildren },
              },
            },
          },
        }
      ) {
        // Update apollo cache after successful db update
        cache.writeQuery({
          query: GET_FAMILY_CHILD_RECIEPIENT_CACHE,
          data: {
            recipient: { children: updatedChildren, __typename: 'Recipient' },
          },
          variables: {
            uuid: match.params.id,
          },
        });
      },
    }
  );

  const [updateRecipient, { error }] = useMutation(UPDATE_RECIPIENT, {
    update(
      cache,
      {
        data: {
          updateRecipient: { recipient },
        },
      }
    ) {
      cache.writeQuery({
        query: GET_FAMILY_RECIEPIENT_CACHE,
        data: {
          recipient,
          __typename: 'Recipient',
        },
        variables: {
          uuid: match.params.id,
        },
      });
    },
  });

  const [recipientData, setRecipientData] = useState({
    currentValue: allFieldsCurrentValue,
    items: [...allOptions],
    photos: allPhotoFields,
    dates: allDateFields,
    times: allTimeFields,
  });

  const [expanded, setExpanded] = useState(false);

  const reArrangeFields = function(arr) {
    const from = arr.indexOf(arr.find(a => a.props.type === 'photo'));
    const to = arr.length - 1;
    arr.splice(to, 0, arr.splice(from, 1)[0]);
  };

  const handleOptionChange = (value, field) => {
    setRecipientData(state => ({
      ...state,
      currentValue: { ...state.currentValue, [field]: { item: value } },
    }));
  };

  const handleDateTimeChange = (value, field, types) => {
    setRecipientData(state => ({
      ...state,
      [types]: { ...state[types], [field]: value },
    }));
  };

  if (customFields) {
    customFields
      .filter((f, i) => {
        if (f.option_values) {
          if (f.field_type === 'multi_select') {
            allFieldsCurrentValue[f.name] = {
              item: !data.customFields[f.name] ? [] : data.customFields[f.name],
            };
          } else {
            allFieldsCurrentValue[f.name] = {
              item: !data.customFields[f.name]
                ? f.option_values[0]
                : data.customFields[f.name],
            };
          }
        }
        if (f.field_type === 'photo') {
          allPhotoFields[f.name] = data.customFields[f.name];
        }

        if (f.field_type === 'date') {
          allDateFields[f.name] = data.customFields[f.name];
        }

        if (f.field_type === 'time') {
          allTimeFields[f.name] = data.customFields[f.name];
        }

        singleFields.push(
          ftSwitch(
            f,
            data,
            i,
            recipientData,
            handleOptionChange,
            handleDateTimeChange,
            parent ? updateRecipient : updateRecipientChild
          )
        );
        return f.option_values;
      })
      .map(o => o.option_values)
      .forEach(group => {
        group.forEach(o => {
          allOptions = [...allOptions, { value: o }];
        });
      });

    reArrangeFields(singleFields);

    if (twoColumn && customFields.length > 0) {
      for (let i = 0; i < singleFields.length; i += 2) {
        const current = singleFields[i];
        const next = singleFields[i + 1];

        twoColfields = [
          ...twoColfields,
          <TwoColumns
            key={i + 5 * 10}
            col1={current || <></>}
            col2={next || <></>}
          />,
        ];
      }
    }
  }

  if (error || childError) {
    return generalErrorHandler(error || childError, { ...rest });
  }

  return (
    <>
      <ExpansionPanel onChange={(event, open) => setExpanded(open)}>
        <ExpansionPanelSummary
          expandIcon={<SmallRightArrow />}
          aria-controls="family-member"
          id={name.toLowerCase().replace(/ /g, '')}
        >
          <Grid container direction="column" justify="flex-start">
            <Typography variant="body2" className={classes.contentBody}>
              {name}
            </Typography>
            {type && (
              <Typography variant="caption" className={classes.caption}>
                {type}
              </Typography>
            )}
          </Grid>
        </ExpansionPanelSummary>

        <ExpansionPanelDetailsStyle>
          <ExpansionPanelDetails>
            <Grid container direction="row">
              <Grid item xs={6}>
                <CardMedia
                  className={classes.recipientChildPhoto}
                  component="img"
                  alt={data.name}
                  width="100%"
                  image={
                    data.base64Avatar
                      ? generateImg(data.base64Avatar)
                      : placeHolderImg
                  }
                  title={data.name}
                />
              </Grid>
              <Grid item xs={12}>
                {twoColumn ? twoColfields : singleFields}

                <Divider light />

                {enableDelete && (
                  <Grid
                    item
                    container
                    direction="row"
                    justify="flex-end"
                    alignItems="flex-end"
                    className={rsClasses.deleteRecipientContainer}
                  >
                    <NSButton
                      textBtn
                      variant="contained"
                      size="small"
                      onClick={() => console.log('DELETING MEMBER')}
                      componentClasses="recipient-delete-btn"
                    >
                      <DeleteIcon className={rsClasses.deleteIcon} />

                      <Typography
                        variant="body1"
                        className={classes.recipientDeleteBtnText}
                      >
                        {intl.formatMessage(globalMessages.deleteRecipient)}
                      </Typography>
                    </NSButton>
                  </Grid>
                )}

                {!lastChild && <Divider className={rsClasses.divider} />}
              </Grid>
            </Grid>
          </ExpansionPanelDetails>
        </ExpansionPanelDetailsStyle>
      </ExpansionPanel>
      {!lastChild && !expanded && <Divider light className={classes.divider} />}
    </>
  );
}

RecipientPanel.propTypes = {
  intl: intlShape.isRequired,
  name: PropTypes.string.isRequired,
  customFields: PropTypes.array,
  data: PropTypes.object.isRequired,
  lastChild: PropTypes.bool,
  parent: PropTypes.bool,
  twoColumn: PropTypes.bool,
  enableDelete: PropTypes.bool,
  type: PropTypes.string,
};

const intlRecipientPanel = injectIntl(RecipientPanel);

export default intlRecipientPanel;
