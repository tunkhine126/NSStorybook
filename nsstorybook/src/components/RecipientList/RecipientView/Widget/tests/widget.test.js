/**
 *
 * Tests for Widget
 *
 * @see https://github.com/react-boilerplate/react-boilerplate/tree/master/docs/testing
 *
 */

import React from 'react';
import TestingComponent from 'components/shared/Testing';
import { render } from '@testing-library/react';
// import 'jest-dom/extend-expect'; // add some helpful assertions
import Widget from '../widget';

describe('<Widget />', () => {
  it('Expect to not log errors in console', () => {
    const spy = jest.spyOn(global.console, 'error');
    render(
      <TestingComponent>
        <Widget
          header="Test Header"
          body={<div>Test Content</div>}
          message="addRecipientChild"
          btnHandler={() => {}}
        />
      </TestingComponent>
    );
    expect(spy).not.toHaveBeenCalled();
  });

  it('Should render and match the snapshot', () => {
    const {
      container: { firstChild },
    } = render(
      <TestingComponent>
        <Widget
          header="Test Header"
          body={<div>Test Content</div>}
          message="addRecipientChild"
          btnHandler={() => {}}
        />
      </TestingComponent>
    );
    expect(firstChild).toMatchSnapshot();
  });
});
