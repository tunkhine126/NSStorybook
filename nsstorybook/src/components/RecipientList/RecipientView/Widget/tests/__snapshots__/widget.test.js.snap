// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Widget /> Should render and match the snapshot 1`] = `
<div
  class="MuiPaper-root makeStyles-paper-313 MuiPaper-elevation1 MuiPaper-rounded"
>
  <p
    class="MuiTypography-root makeStyles-widgetHeader-324 makeStyles-widgetPadding-325 MuiTypography-body1 MuiTypography-gutterBottom"
  >
    Test Header
  </p>
  <hr
    class="MuiDivider-root MuiDivider-light"
  />
  <div
    class="styles__ExpansionPanelStyle-qa8k4k-9 iyqZhl"
  >
    <div>
      Test Content
    </div>
    <hr
      class="MuiDivider-root MuiDivider-light"
    />
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-direction-xs-column MuiGrid-align-items-xs-flex-start MuiGrid-justify-xs-center"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text makeStyles-addBtn-273 MuiButton-textPrimary makeStyles-textPrimary-498"
        style="margin-right: auto; pointer-events: initial;"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label makeStyles-label-497"
        >
          <span>
            Add Recipient
          </span>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</div>
`;
